packages/design/src/theme/theme.tsx

```diff
@@ -126,8 +198,6 @@ const theme = {
   breakpoints: [],
 };
```

Comment:
Alternatively we can infer theme type from value:
```tsx
const space = [0, 4, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80] as const;
const contrastThreshold = 3;

export type PropsWithTheme<T> = { theme: Theme } & T;

const colors = {
  accent: '#651FFF',

  dark: '#000',

  light: '#FFFFFF',

  primary: {
    main: '#1C254D',
    light: '#222C59',
    lighter: '#2C3A73',
    dark: '#111B48',
    contrastText: '#FFFFFF',
  },

  secondary: {
    main: '#512FC9',
    light: '#651FFF',
    dark: '#354AA4',
    contrastText: '#FFFFFF',
  },

  text: {
    // The most important text.
    primary: 'rgba(255,255,255,0.87)',
    // Secondary text.
    secondary: 'rgba(255, 255, 255, 0.56)',
    // Placeholder text for forms.
    placeholder: 'rgba(255, 255, 255, 0.24)',
    // Disabled text have even lower visual prominence.
    disabled: 'rgba(0, 0, 0, 0.24)',
    // Text hints.
    hint: 'rgba(0, 0, 0, 0.24)',
    // On light backgrounds
    onLight: '#324148',
    // On dark backgrounds
    onDark: 'rgba(255, 255, 255, 0.87)',
  },

  grey: blueGrey,

  error: {
    light: red['A200'],
    dark: red['A700'],
    main: red['A400'],
  },

  action: {
    active: '#FFFFFF',
    hover: 'rgba(255, 255, 255, 0.1)',
    hoverOpacity: 0.1,
    selected: 'rgba(255, 255, 255, 0.2)',
    disabled: 'rgba(255, 255, 255, 0.3)',
    disabledBackground: 'rgba(255, 255, 255, 0.12)',
  },

  subtle: blueGrey[50],
  link: lightBlue[500],
  bgTerminal: '#010B1C',
  danger: pink.A400,
  highlight: yellow[50],
  disabled: blueGrey[500],
  info: lightBlue[600],
  warning: orange.A400,
  success: teal.A700,
} as const;

const borders = [
  0,
  '1px solid',
  '2px solid',
  '4px solid',
  '8px solid',
  '16px solid',
  '32px solid',
] as const;

export const theme = {
  colors,
  typography,
  font: fonts.sansSerif,
  fonts: fonts,
  fontWeights,
  fontSizes,
  space,
  borders,
  radii: [0, 2, 4, 8, 16, 9999, '100%'],
  regular: fontWeights.regular,
  bold: fontWeights.bold,
  // disabled media queries for styled-system
  breakpoints: [],
} as const;

export type Theme = typeof theme;
```

Reply:
I'd like to, but they differ slightly between enterprise and OSS.

To be honest, I hate the entire theming code and think we should refactor this in a PR afterwards - thoughts?

Reply:
@ryanclark this PR looks too complex for a relatively simple problem. For example, you can look at the `next` (or website) repo how it's done there. Basically the latest styled-system provides a good typescript support. You can start refactoring the code to match the latest styled-system (theme and lookup methods), unifying OSS/E themes and the rest will come alone. If that for some reason is still not enough then building a custom theme would be an option but merging this PR is not great as this is a lot of [unmaintainable code](https://github.com/gravitational/webapps/blob/ryan/type-the-design-system/packages/design/src/styled/styled.ts) that creates more issues than solves.

Reply:
This is also introducing automatic typing for composed styled elements from `styled-system`, without having to keep the typed props and the usages inline. This is taking advantage of the latest `styled-systems` typings.

I don't see how the code you linked is unmaintainable, as it is from styled-components typings themselves and adapted to work for us. If they update their types, we can do the same.

Reply:
Yeah, my concern that it relies on internal styled-component implementation and that maintaining it is now Teleport responsibility. Which is both risky and time consuming as it's hard to maintain something you have no control over. Even a question "Is it styled component bug/limitation or ours" is no longer a simple answer.  But I appreciate the intent, i just hopped that it could've been done with less hacky approach.

Reply:
I'm a bit unsure about `styled.ts` too :( 
If I would have to update this file I wouldn't know what exactly should I do. What part of it is just copied from `styled-components` and what is changed to work for us?

Also, it would be worth to leave a comment why we need to do this.

Reply:
I wrote a doc about what this brings us here - https://github.com/gravitational/webapps/blob/ryan/type-the-design-system/docs/styled-components.md

I can document `styled.ts` to show exactly what is ours vs theirs, to make any changes easier on future developers. I get that it looks like a large scary typescript file, but most of it is just copied from `@types/styled-components` and adapted where we need a different constructor, or type inference.

Reply:
>I can document `styled.ts` to show exactly what is ours vs theirs, to make any changes easier on future developers.

Yeah, I think it would be helpful.

Reply:
Done! I've added comments about what we've copied, why we copied specific things as well as what we've added and what was their original typings.